### PR TITLE
[19.09] Fallback to binary download if header lies about content encoding

### PR DESCRIPTION
**Pull request by mvdbeek**
Fixes https://github.com/galaxyproject/galaxy/issues/9238

### CODE REVIEW

**Comment by mvdbeek** [comment]
The integration test failure is unrelated.

File: lib/galaxy/datatypes/sniff.py [hunk 1]
```diff
@@ -64,15 +64,20 @@ def stream_to_open_named_file(stream, fd, filename, source_encoding=None, source
         codecs.lookup(target_encoding)
     except Exception:
         target_encoding = util.DEFAULT_ENCODING  # utf-8
+    try_source_encoding = True
```
Comment (nsoranzo) [inline lib/galaxy/datatypes/sniff.py]:
```suggestion
    use_source_encoding = source_encoding is not None
```

File: lib/galaxy/datatypes/sniff.py [hunk 2]
```diff
@@ -64,15 +64,20 @@ def stream_to_open_named_file(stream, fd, filename, source_encoding=None, source
...
     while True:
         chunk = stream.read(CHUNK_SIZE)
         if not chunk:
             break
-        if source_encoding is not None:
+        if source_encoding is not None and try_source_encoding:
```
Comment (nsoranzo) [inline lib/galaxy/datatypes/sniff.py]:
```suggestion
        if use_source_encoding:
```

File: lib/galaxy/datatypes/sniff.py [hunk 3]
```diff
@@ -64,15 +64,20 @@ def stream_to_open_named_file(stream, fd, filename, source_encoding=None, source
...
             # If a source encoding is given we use it to convert to the target encoding
-            if not isinstance(chunk, text_type):
-                chunk = chunk.decode(source_encoding, source_error)
-            os.write(fd, chunk.encode(target_encoding, target_error))
+            try:
+                if not isinstance(chunk, text_type):
+                    chunk = chunk.decode(source_encoding, source_error)
+                os.write(fd, chunk.encode(target_encoding, target_error))
+            except UnicodeDecodeError:
+                try_source_encoding = False
```
Comment (nsoranzo) [inline lib/galaxy/datatypes/sniff.py]:
```suggestion
                use_source_encoding = False
```

**Review by nsoranzo — APPROVED** [review]

**Pull request merged**